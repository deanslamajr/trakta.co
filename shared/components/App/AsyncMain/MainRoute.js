import React from 'react';
import Helmet from 'react-helmet';
import withStyles from 'isomorphic-style-loader/lib/withStyles';
import axios from 'axios';

import config from '../../../../config';

import Recorder from '../../../../client/components/Recorder';
import SampleInstances from '../../../../client/components/SampleInstances';
import InstancePlaylist from '../../../../client/components/InstancePlaylist';

import styles from './styles.css';

// @todo dynamically generate these
const windowStartTime = 0;
const windowLength = 5;

class MainRoute extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isClient: false,
      subview: null,
      instances: []
    };

    this._showContribute = this._showContribute.bind(this);
    this._showMainMenu = this._showMainMenu.bind(this);
    this._renderLoadingComponent = this._renderLoadingComponent.bind(this);
    this._renderPlayComponent = this._renderPlayComponent.bind(this);

    this.views = {
      contribute: Recorder
    }
  }

  /**
   * Retrieve sample instances for current viewport
   */
  _getSampleInstances() {
    axios.get('/api/sampleInstances')
      .then(({ data }) => {
        this.setState({
          instances: data
        })
      })
      .catch(error => {
        // @todo log
        console.error(error);
      });
  }

  _showContribute() {
    this.setState({
      subview: this.views.contribute
    })
  }

  _renderLoadingComponent() {
    return (<div className={styles.loadingButton}>Loading</div>);
  }

  _renderPlayComponent(clickHandler) {
    return (<div className={styles.playButton} onClick={clickHandler}>&#128266;</div>);
  }

  _renderMainMenu() {
    // on server, this should only concern itself with displaying a load animation
    // on top of notched track background, showing the correct time labels related to
    // the current track viewport
    return (
      <div className={styles.canvasContainer}>
        <div className={styles.label}>
          <InstancePlaylist
            instances={this.state.instances}
            renderLoadingComponent={this._renderLoadingComponent}
            renderPlayButtonComponent={this._renderPlayComponent}
            windowLength={windowLength} 
            windowStartTime={windowStartTime} />
          <div className={styles.contributeButton} onClick={this._showContribute}>&#10133;</div>
        </div>
        
        { this.state.isClient &&
            <SampleInstances 
              instances={this.state.instances}
              windowLength={windowLength} 
              windowStartTime={windowStartTime}/>
        }
      </div>
    )
  }

  _showMainMenu() {
    this._getSampleInstances();

    this.setState({
      subview: null
    })
  }

  componentDidMount() {
    this.setState({ isClient: true });
    this._getSampleInstances();
  }

  render() {
    const Subview = this.state.subview;

    return (
      <div className={styles.container}>
        <Helmet>
          <title>{`recorder - ${config('appTitle')}`}</title>
        </Helmet>
        { Subview
            ? <Subview showMainMenu={this._showMainMenu} />
            : this._renderMainMenu()
        }
      </div>
    );
  }
}

export default withStyles(styles)(MainRoute)

export { MainRoute }