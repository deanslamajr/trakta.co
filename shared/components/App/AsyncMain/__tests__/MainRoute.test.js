/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import { shallow } from 'enzyme';

import { MainRoute } from '../MainRoute';

describe('<MainRoute />', () => {
  test('renders', () => {
    const wrapper = shallow(<MainRoute />);
    expect(wrapper).toMatchSnapshot();
  });
});
