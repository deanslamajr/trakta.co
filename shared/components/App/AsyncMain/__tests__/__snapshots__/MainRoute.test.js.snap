// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainRoute /> renders 1`] = `
<div
  className="container"
>
  <HelmetWrapper
    encodeSpecialCharacters={true}
  >
    <title>
      recorder - tracta.co
    </title>
  </HelmetWrapper>
  <div
    className="canvasContainer"
  >
    <div
      className="meter"
    >
      <span
        className="startTime"
      >
        0
      </span>
      <span
        className="endTime"
      >
        20
      </span>
    </div>
    <div
      className="label"
    >
      <AsyncComponent
        renderErrorComponent={[Function]}
        windowLength={20}
        windowStartTime={0}
      />
      <div
        className="contribute button bottomButton"
        onClick={[Function]}
      >
        <span
          className="icon"
        >
          âž•
        </span>
      </div>
    </div>
    <AsyncComponent
      windowLength={20}
      windowStartTime={0}
    />
  </div>
</div>
`;
